File: PythonSpark/Aggregations/AggregationJob.py
from abc import ABC

from pyspark.sql import DataFrame

from JobManagment.Job import Job
from Utils.RichArgumentParser import RichArgumentParser
from Utils.Writing import write_to_postgres, write_to_big_query


class AggregationJob(Job, ABC):
    def write(self, data: DataFrame) -> None:
        if self.arguments.mode == "dev":
            write_to_postgres(data,
                              self.arguments.postgres_url,
                              self.arguments.postgres_user,
                              self.arguments.postgres_password,
                              self.arguments.result_table)
        else:
            write_to_big_query(data,
                               self.arguments.big_query_service_account_key_filepath,
                               self.arguments.temporary_bucket_name,
                               self.arguments.result_table)

    @staticmethod
    def parser() -> RichArgumentParser:
        return Job.parser() \
            .add_postgres() \
            .add_big_query() \
            .add_result_table()

File: ScalaSpark/src/scala/Aggregations/AggregationJob.scala
package Aggregations

import JobManagment.{Job, JobCompanion}
import Utils.ArgumentsParsing.RichParser
import Utils.Writing.{writeToBigQuery, writeToPostgres}
import de.halcony.argparse.Parser
import org.apache.spark.sql.{DataFrame, SparkSession}

abstract class AggregationJob(arguments: Map[String, String], spark: SparkSession) extends Job(arguments, spark) {
  override def write(data: DataFrame): Unit = {
    arguments("mode") match {
      case "dev" => writeToPostgres(arguments("postgresUrl"), arguments("resultTable"), arguments("postgresUser"), arguments("postgresPassword"), data)
      case "prod" => writeToBigQuery(arguments("temporaryBucketName"), arguments("resultTable"), arguments("bigQueryServiceAccountKeyFilepath"), data)
    }
  }
}


object AggregationJob extends JobCompanion {
  override def parser(): Parser = {
    Job.parser()
      .addMode()
      .addPostgres()
      .addBigQuery()
      .addResultTable()
  }
}